File: 20.valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # stack_braces = []
        # stack_brackets = []
        # stack_parentheses = []

        # for char in s:
        #     if char == "(":
        #         stack_parentheses.append(char)
        #     if char == "[":
        #         stack_brackets.append(char)
        #     if char == "{":
        #         stack_braces.append(char)
        #     if char == ")":
        #         if len(stack_parentheses) == 0:
        #             return False
        #         stack_parentheses.pop()
        #     if char == "]":
        #         if len(stack_brackets) == 0:
        #             return False
        #         stack_brackets.pop()
        #     if char == "}":
        #         if len(stack_braces) == 0:
        #             return False
        #         stack_braces.pop()

        # return stack_braces == [] and stack_brackets == [] and stack_parentheses == []

        stack = []

        for c in s:
            if c == "(":
                stack.append(")")
            elif c == "[":
                stack.append("]")
            elif c == "{":
                stack.append("}")
            
            if c == ")":
                if len(stack) == 0:
                    return False
                if stack.pop() != ")":
                    return False
            
            if c == "]":
                if len(stack) == 0:
                    return False
                if stack.pop() != "]":
                    return False
            
            if c == "}":
                if len(stack) == 0:
                    return False
                if stack.pop() != "}":
                    return False
        
        return len(stack) == 0

if __name__ == "__main__":
    solution = Solution()
    
    string = "{[()}"

    print(solution.isValid(string))

File: 20.valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  if (s.length % 2 === 1) return false;

  const arr = s.split("");

  const map = new Map();
  map.set("(", ")");
  map.set("{", "}");
  map.set("[", "]");

  const stack = [];

  for (let i = 0; i < arr.length; i++) {
    if (map.has(arr[i])) {
      stack.push(map.get(arr[i]));
    } else {
      if (arr[i] === stack[stack.length - 1]) stack.pop();
      else return false;
    }
  }
  if (stack.length > 0) return false;
  return true;
};
